### PR TITLE
feat: add onboard screens

### LINKED ISSUE #11: Implement onboarding landing page
Implement the onboarding landing page as a fragment in the Onboarding Activity.

Please define related colors and text & button styles in the corresponding values files.

Some basic scaffolding was added, but it needs to be cleaned up. 

image

[Link to design in Figma](https://www.figma.com/file/d6n3MPVRWbL5jP2NmvW96b/De_Urgenta?node-id=1466%3A530)

### PR DESCRIPTION
**Pull request by toaderandrei**
* Add onboard screens to configure user profile,
ConfigureAddressFragment, ConfigureAccountFragment and their respective
view models.
* Add nav_graph for onboard screens.
* Add class to bind views from layout to objects.
* Add MapViewUtils to handle map related stuff.
* Add PermissionUtils to handle permissions related stuff.
* Add callbacks for click button events.
* Add google maps for localization.
* Add temporary icons used for the layouts.
* Add resources values as well.

Implements: #11, #13 

### What does it fix?

Closes #11, #13

### CODE REVIEW

File: app/src/main/java/ro/code4/deurgenta/bindings/BindingAdapters.kt [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+import android.view.ViewGroup
+import android.widget.ImageView
+import androidx.databinding.BindingAdapter
+import com.bumptech.glide.Glide
+
+object BindingAdapters {
+    @SuppressLint("ResourceType")
+    @JvmStatic
+    @BindingAdapter(value = ["imageRes"], requireAll = false)
+    fun bindImageResource(imageView: ImageView, resId: Int?) {
```
Comment (aniri) [inline app/src/main/java/ro/code4/deurgenta/bindings/BindingAdapters.kt]: where is this used?

File: app/src/main/AndroidManifest.xml [hunk 2]
```diff
@@ -4,6 +4,17 @@
...
+
+
+    <uses-permission android:name="android.permission.ACCESS_COARSE_LOCATION" />
+    <uses-permission android:name="android.permission.ACCESS_FINE_LOCATION" />
+    <uses-permission android:name="android.permission.BLUETOOTH" />
+    <uses-permission android:name="android.permission.BLUETOOTH_ADMIN"/>
+    <uses-permission android:name="android.permission.CHANGE_NETWORK_STATE" />
+    <uses-permission android:name="android.permission.CHANGE_WIFI_STATE" />
+    <uses-permission android:name="android.permission.READ_PHONE_STATE" android:maxSdkVersion="22" />
+    <uses-permission android:name="android.permission.WAKE_LOCK" />
```
Comment (aniri) [inline app/src/main/AndroidManifest.xml]: please only add needed permissions

Comment (toaderandrei) [reply app/src/main/AndroidManifest.xml]: Ofc. Only needed permissions. This is the amount of permissions here sdk requires, at least in their sample apps. I will try to do some tests, maybe some can be removed. However, It might that some are specific to different android api levels.

File: app/src/main/java/ro/code4/deurgenta/helper/MapViewUtils.kt [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+        clearMap()
+        if (queryString.length < 2) {
+            return
+        }
+
+        val viewportGeoBox: GeoBox = getMapViewGeoBox()
+        val query = TextQuery(queryString, viewportGeoBox)
+
+        val maxItems = DEFAULT_MAX_ITEMS
+        val searchOptions = SearchOptions(LanguageCode.EN_GB, maxItems)
```
Comment (aniri) [inline app/src/main/java/ro/code4/deurgenta/helper/MapViewUtils.kt]: you could use DEFAULT_MAX_ITEMS directly here, no need for extra var

also, shouldn't the search language be set to ro?

File: app/src/main/java/ro/code4/deurgenta/bindings/BindingAdapters.kt (hunk 1 above)
Comment (toaderandrei) [reply app/src/main/java/ro/code4/deurgenta/bindings/BindingAdapters.kt]: Correct, not used atm. It was used but then I decided to do it in the next PR. Will remove it.

File: app/src/main/java/ro/code4/deurgenta/helper/MapViewUtils.kt [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+        searchEngine.search(query, searchOptions, querySearchCallback)
+    }
+
+    fun searchOnMapAutoSuggestion(queryString: String) {
```
Comment (aniri) [inline app/src/main/java/ro/code4/deurgenta/helper/MapViewUtils.kt]: almost 100% duplicated code with method from above

File: app/src/main/java/ro/code4/deurgenta/helper/MapViewUtils.kt (hunk 3 above)
Comment (toaderandrei) [reply app/src/main/java/ro/code4/deurgenta/helper/MapViewUtils.kt]: Probably, but not sure it actually matters. I am searching in Austria, no difference seen. I will do some tests with country related code.

File: app/src/main/java/ro/code4/deurgenta/ui/address/ConfigureAddressFragment.kt [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+        mapViewUtils.searchOnMap(queryString)
+    }
+
+    private fun searchOnMapAutoSuggestion(queryString: String) {
+        mapViewUtils.searchOnMapAutoSuggestion(queryString)
+    }
+
+    fun dismissKeyboard(view: View) {
+        hideSoftInput(view)
+    }
```
Comment (aniri) [inline app/src/main/java/ro/code4/deurgenta/ui/address/ConfigureAddressFragment.kt]: what is the utility of these methods?

Comment (toaderandrei) [reply app/src/main/java/ro/code4/deurgenta/ui/address/ConfigureAddressFragment.kt]: Basically it is about the hiding of the keyboard when you start the fragment for the first time.

File: app/src/main/java/ro/code4/deurgenta/ui/base/Layout.kt [hunk 6]
```diff
@@ -0,0 +1,10 @@
+package ro.code4.deurgenta.ui.base
+
+import androidx.annotation.LayoutRes
+import ro.code4.deurgenta.interfaces.ExcludeFromCodeCoverage
+
+interface Layout {
```
Comment (aniri) [inline app/src/main/java/ro/code4/deurgenta/ui/base/Layout.kt:6]: this interface already existed

File: app/src/main/java/ro/code4/deurgenta/ui/base/ViewModelSetter.kt [hunk 7]
```diff
@@ -0,0 +1,8 @@
+package ro.code4.deurgenta.ui.base
+import androidx.lifecycle.ViewModel
+import ro.code4.deurgenta.interfaces.ExcludeFromCodeCoverage
+
+interface ViewModelSetter<out T : ViewModel> {
```
Comment (aniri) [inline app/src/main/java/ro/code4/deurgenta/ui/base/ViewModelSetter.kt:5]: also duplicated

File: app/src/main/res/drawable/roundedbutton.xml [hunk 8]
```diff
@@ -0,0 +1,9 @@
+<?xml version="1.0" encoding="utf-8"?>
+<shape xmlns:android="http://schemas.android.com/apk/res/android"
```
Comment (aniri) [inline app/src/main/res/drawable/roundedbutton.xml]: could we reuse the buttons from auth activity?

File: app/src/main/res/layout/onboarding_configure_account.xml [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+            android:layout_marginTop="@dimen/xxbig_margin"
+            android:orientation="vertical"
+            app:layout_constraintTop_toBottomOf="@+id/tv_configure_account_header">
+
+            <TextView
+                android:id="@+id/configure_address"
+                style="@style/MenuTextView"
+                android:background="?android:attr/selectableItemBackground"
+                android:onClick="@{() -> callbackAddresses.configure()}"
+                android:padding="12dp"
```
Comment (aniri) [inline app/src/main/res/layout/onboarding_configure_account.xml]: please use dimens

File: app/src/main/res/layout/onboarding_configure_addresses.xml [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+            android:fitsSystemWindows="true"
+            android:theme="@style/AppTheme">
+
+            <com.google.android.material.appbar.MaterialToolbar
+                android:id="@+id/toolbar"
+                android:layout_width="match_parent"
+                android:layout_height="?attr/actionBarSize"
+                android:background="@android:color/transparent"
+                app:navigationIcon="@drawable/ic_close_24"
+                app:title="Add your own address"
```
Comment (aniri) [inline app/src/main/res/layout/onboarding_configure_addresses.xml]: extract to strings

File: app/src/main/res/layout/onboarding_configure_addresses.xml [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+        <androidx.constraintlayout.widget.ConstraintLayout
+            android:layout_width="match_parent"
+            android:layout_height="match_parent"
+            app:layout_behavior="@string/appbar_scrolling_view_behavior">
+
+            <androidx.appcompat.widget.AppCompatButton
+                android:id="@+id/bt_locate_me"
+                android:layout_width="wrap_content"
+                android:layout_height="wrap_content"
+                android:layout_marginTop="25dp"
```
Comment (aniri) [inline app/src/main/res/layout/onboarding_configure_addresses.xml]: dimens

**Comment by aniri** [comment]
the app crashed when I started to write an address in the input field :( 
```
E/AndroidRuntime: FATAL EXCEPTION: main
    Process: ro.code4.deurgenta, PID: 7850
    java.lang.RuntimeException: GeoBox creation failed, corners are null.
        at ro.code4.deurgenta.helper.MapViewUtils.getMapViewGeoBox(MapViewUtils.kt:61)
        at ro.code4.deurgenta.helper.MapViewUtils.searchOnMapAutoSuggestion(MapViewUtils.kt:88)
        at ro.code4.deurgenta.ui.address.ConfigureAddressFragment.searchOnMapAutoSuggestion(ConfigureAddressFragment.kt:188)
        at ro.code4.deurgenta.ui.address.ConfigureAddressFragment.access$searchOnMapAutoSuggestion(ConfigureAddressFragment.kt:30)
        at ro.code4.deurgenta.ui.address.ConfigureAddressFragment$onQueryTextListener$1.onQueryTextChange(ConfigureAddressFragment.kt:160)
```

**Comment by toaderandrei** [comment]
It could if the map was not initialized somehow. Will check this.

File: app/src/main/res/values/dimens.xml [hunk 12]
```diff
@@ -1,13 +1,31 @@
 <resources>
```
Comment (aniri) [inline app/src/main/res/values/dimens.xml]: please refrain from adding too many different values and follow material guidelines https://material.io/archive/guidelines/layout/metrics-keylines.html#metrics-keylines-keylines-spacing
the layout doesn't have to be 100% aligned with the design

File: app/src/main/java/ro/code4/deurgenta/ui/address/ConfigureAddressFragment.kt (hunk 5 above)
Comment (aniri) [reply app/src/main/java/ro/code4/deurgenta/ui/address/ConfigureAddressFragment.kt]: I meant why do we need these 3 methods since we could just use the methods from maputils/utils directly. the method from this class do not bring extra functionality

File: app/src/main/java/ro/code4/deurgenta/interfaces/ConfigureCallback.kt [hunk 13]
```diff
@@ -0,0 +1,5 @@
+package ro.code4.deurgenta.interfaces
+
+interface ConfigureCallback {
```
Comment (aniri) [inline app/src/main/java/ro/code4/deurgenta/interfaces/ConfigureCallback.kt:3]: please rename this to something more clear, maybe configureaddresscallback?

File: app/src/main/res/layout/onboarding_configure_account.xml [hunk 14]
```diff
@@ -0,0 +1,110 @@
...
+                android:text="@string/create_your_group"
+                android:textAppearance="@style/Text.OptionItem"
+                app:drawableStartCompat="@drawable/ic_group" />
+
+            <TextView
+                android:id="@+id/make_backpack"
+                style="@style/MenuTextView"
+                android:background="?android:attr/selectableItemBackground"
+
+                android:padding="12dp"
```
Comment (aniri) [inline app/src/main/res/layout/onboarding_configure_account.xml]: please reuse values from dimens

**Comment by aniri** [comment]
the app still crashes with the same error:
```
D/AndroidRuntime: Shutting down VM
E/AndroidRuntime: FATAL EXCEPTION: main
    Process: ro.code4.deurgenta, PID: 4731
    java.lang.RuntimeException: GeoBox creation failed, corners are null.
        at ro.code4.deurgenta.helper.MapViewUtils.getMapViewGeoBox(MapViewUtils.kt:61)
        at ro.code4.deurgenta.helper.MapViewUtils.searchOnMap(MapViewUtils.kt:78)
        at ro.code4.deurgenta.helper.MapViewUtils.searchOnMap$default(MapViewUtils.kt:71)
        at ro.code4.deurgenta.ui.address.ConfigureAddressFragment$onQueryTextListener$1.onQueryTextChange(ConfigureAddressFragment.kt:165)
```

both on the phone and emulator.

could you please investigate?

File: app/src/main/java/ro/code4/deurgenta/helper/MapViewUtils.kt [hunk 15]
```diff
@@ -0,0 +1,275 @@
...
+                    mapView.camera.lookAt(geoCoordinates, distanceInMeters)
+                    lastGeoCoordinates = geoCoordinates
+                    areLocationServicesInitialized = true
+                    if (showLocation) {
+                        createAndAddMarker(geoCoordinates)
+                    }
+                } else {
+                    Log.e(TAG, "error loading results:$errorCode")
+                    if (errorCode != MapError.OPERATION_IN_PROGRESS) {
+                        callback.onError(MapDataError("cannot load map. Error:$errorCode", MapErrorType.NOT_INITIALIZED))
```
Comment (aniri) [inline app/src/main/java/ro/code4/deurgenta/helper/MapViewUtils.kt]: please extract error strings in strings.xml

File: app/src/main/java/ro/code4/deurgenta/helper/MapViewUtils.kt [hunk 16]
```diff
@@ -0,0 +1,275 @@
...
+            val searchOptions = SearchOptions(LanguageCode.RO_RO, maxSearchResults)
+
+            if (searchType == SearchType.STANDARD) {
+                searchEngine.search(query, searchOptions, querySearchCallback)
+            } else {
+                searchEngine.suggest(query, searchOptions, autosuggestCallback)
+            }
+        } else {
+            if (searchType == SearchType.STANDARD) {
+                callback.onError(MapDataError("Map is not initialized", MapErrorType.NOT_INITIALIZED));
```
Comment (aniri) [inline app/src/main/java/ro/code4/deurgenta/helper/MapViewUtils.kt]: please extract error strings in strings.xml

File: app/src/main/java/ro/code4/deurgenta/helper/MapViewUtils.kt [hunk 17]
```diff
@@ -0,0 +1,275 @@
...
+                Log.e(TAG, "fail to get a response.$e")
+
+                when ((e as ApiException).statusCode) {
+                    LocationSettingsStatusCodes.RESOLUTION_REQUIRED -> {
+                        callback.onError(MapDataError("needs more permissions", MapErrorType.PERMISSION_ERROR, e))
+                        Log.i(TAG, "Location settings are not satisfied. Attempting to upgrade location settings.")
+
+                    }
+                    LocationSettingsStatusCodes.SETTINGS_CHANGE_UNAVAILABLE -> {
+                        val errorMessage = "Location settings are inadequate, and cannot be fixed here. Fix in Settings."
```
Comment (aniri) [inline app/src/main/java/ro/code4/deurgenta/helper/MapViewUtils.kt]: please extract error strings in strings.xml

**Comment by aniri** [comment]
please extract all error texts that might be displayed to the users in the strings.xml files

File: app/build.gradle [hunk 18]
```diff
@@ -93,24 +97,15 @@ android {
...
-    navigationVersion = "2.3.4"
-    koinVersion = '2.0.1'
-    parcelerVersion = "1.1.13"
-    roomVersion = '2.2.6'
-    firebaseAnalyticsVersion = '18.0.2'
-    firebaseCrashlyticsVersion = '17.4.1'
-    firebaseMessagingVersion = "20.2.0"
-    firebaseConfigVersion = "19.1.4"
-    adapterDelegatesVersion = "4.2.0"
+    ndkVersion '20.0.5594570'
```
Comment (aniri) [inline app/build.gradle]: this seems to be added twice and with different versions

File: app/src/main/java/ro/code4/deurgenta/data/AppDatabase.kt [hunk 19]
```diff
@@ -8,16 +8,20 @@ import androidx.room.TypeConverters
...
 import ro.code4.deurgenta.data.model.Backpack
 import ro.code4.deurgenta.data.model.BackpackItem
 import ro.code4.deurgenta.data.model.BackpackItemConverters
+import ro.code4.deurgenta.data.model.MapAddress
 
 @Database(
-    entities = [Address::class, Backpack::class, BackpackItem::class],
-    version = 1
+    entities = [Backpack::class, BackpackItem::class, MapAddress::class],
+    version = 3
```
Comment (aniri) [inline app/src/main/java/ro/code4/deurgenta/data/AppDatabase.kt]: do we really need to increase the version?

File: app/src/main/java/ro/code4/deurgenta/helper/MapViewUtils.kt [hunk 20]
```diff
@@ -0,0 +1,380 @@
...
+
+        val mapMarker = createPoiMapMarker(searchResult.geoCoordinates!!)
+        mapView.mapScene.addMapMarker(mapMarker)
+
+        val mapAddress = MapAddress(
+            latitude = searchResult.geoCoordinates!!.latitude,
+            longitude = searchResult.geoCoordinates!!.longitude,
+            fullAddress = searchResult.address.addressText
+        )
+        mapAddress.streetAddress = getStreetAddress(searchResult.address)
```
Comment (aniri) [inline app/src/main/java/ro/code4/deurgenta/helper/MapViewUtils.kt:135]: you could use formatAddress directly here, no need for another method..

File: app/src/main/java/ro/code4/deurgenta/helper/MapViewUtils.kt [hunk 21]
```diff
@@ -0,0 +1,380 @@
...
+
+        mapMarkerList[mapAddress] = mapMarker
+    }
+
+    private fun getStreetAddress(address: Address): String {
+        return formatAddress(address)
+    }
+
+    private fun getMapViewGeoCircle(): GeoCircle {
+        return GeoCircle(geoCoordinates, 50000.0)
```
Comment (aniri) [inline app/src/main/java/ro/code4/deurgenta/helper/MapViewUtils.kt]: could you extract the radius value in the config files so it could be easily changed?

File: app/src/main/java/ro/code4/deurgenta/helper/MapViewUtils.kt [hunk 22]
```diff
@@ -0,0 +1,380 @@
...
+
+        // Add new marker for each search result on map.
+        for (searchResult in list!!) {
+            createAndAddMarker(searchResult)
+        }
+
+        if (list.size == 1) {
+            loadMapSceneForGeoCoordinates(list[0].geoCoordinates)
+            callback.onSearchSuccess()
+        } else if (list.size > 1) {
```
Comment (aniri) [inline app/src/main/java/ro/code4/deurgenta/helper/MapViewUtils.kt]: I don't think there's a need for an explicit if condition here

File: app/src/main/java/ro/code4/deurgenta/helper/PermissionManager.kt [hunk 23]
```diff
@@ -8,39 +8,34 @@ import androidx.fragment.app.Fragment
...
-                })
+            listener.onPermissionDenied()
             return
         }
-        requestPermissions(*permissionString)
+        checkPermissionsInternal(*permissionString)
     }
 
-    fun requestPermissions(vararg permissionString: String) {
+    private fun checkPermissionsInternal(vararg permissionString: String) {
```
Comment (aniri) [inline app/src/main/java/ro/code4/deurgenta/helper/PermissionManager.kt]: why was this renamed? the new name is not more clear...

File: app/src/main/java/ro/code4/deurgenta/helper/PermissionManager.kt [hunk 24]
```diff
@@ -49,18 +44,23 @@ open class PermissionManager(private val activity: Activity, private val fragmen
...
-        fun onPermissionDenied(vararg allPermissions: String, permissionsDenied: List<String>)
+        fun onPermissionDenied()
     }
+
+    companion object {
+        @JvmStatic
+        val PERMISSION_REQUEST = 320
+
+        @JvmStatic
+        val PERMISSION_CHECK_SETTINGS = 320
```
Comment (aniri) [inline app/src/main/java/ro/code4/deurgenta/helper/PermissionManager.kt]: shouldn't these have different values? 🤔

File: app/src/main/java/ro/code4/deurgenta/interfaces/ViewCallback.kt [hunk 25]
```diff
@@ -0,0 +1,9 @@
+package ro.code4.deurgenta.interfaces
+
+
+/**
+ * Callback for when saving progress.
+ */
+interface ViewCallback {
```
Comment (aniri) [inline app/src/main/java/ro/code4/deurgenta/interfaces/ViewCallback.kt]: could you please rename it to something more clear?

File: app/src/main/res/layout/fragment_configure_account.xml [hunk 26]
```diff
@@ -0,0 +1,117 @@
...
+                android:layout_marginTop="@dimen/xxbig_margin"
+                android:orientation="vertical"
+                app:layout_constraintTop_toBottomOf="@+id/tv_configure_account_header">
+
+                <TextView
+                    android:id="@+id/configure_address"
+                    style="@style/MenuTextView"
+                    android:background="?android:attr/selectableItemBackground"
+                    android:onClick="@{() -> callbackAddresses.call()}"
+                    android:padding="12dp"
```
Comment (aniri) [inline app/src/main/res/layout/fragment_configure_account.xml]: please use values form dimens file

File: app/src/main/res/layout/fragment_configure_account.xml [hunk 27]
```diff
@@ -0,0 +1,117 @@
...
+                    android:text="@string/make_your_emergency_backpack"
+                    android:textAppearance="@style/Text.OptionItem"
+                    app:drawableStartCompat="@drawable/ic_shopping_bag" />
+
+                <TextView
+                    android:id="@+id/learn_firstaid"
+                    style="@style/MenuTextView"
+                    android:background="?android:attr/selectableItemBackground"
+
+                    android:padding="12dp"
```
Comment (aniri) [inline app/src/main/res/layout/fragment_configure_account.xml]: dimens

File: app/src/main/res/values/styles.xml [hunk 28]
```diff
@@ -61,13 +61,26 @@
...
+    </style>
+
     <style name="Text.MenuList">
         <item name="android:textSize">20sp</item>
         <item name="android:textColor">@color/gray_800</item>
         <item name="android:textStyle">bold</item>
         <item name="lineHeight">24sp</item>
     </style>
 
+    <style name="Text.OptionItem">
```
Comment (aniri) [inline app/src/main/res/values/styles.xml]: i think you could have reused Text.MenuList

File: app/src/main/res/values/styles.xml [hunk 29]
```diff
@@ -61,13 +61,26 @@
         <item name="lineHeight">26sp</item>
     </style>
 
+    <style name="Text.Content">
```
Comment (aniri) [inline app/src/main/res/values/styles.xml]: i think you could have reused Text.Default

**Review by aniri — APPROVED** [review]
